tests/rptest/utils/mode_checks.py

```diff
@@ -0,0 +1,24 @@
+import functools
+
+
+def skip_debug_mode(func):
+    @functools.wraps(func)
+    def f(*args, **kwargs):
+        assert args, 'skip_debug_mode must be placed on a test class method'
+
+        caller = args[0]
+
+        assert hasattr(
+            caller, 'debug_mode'
+        ), 'skip_debug_mode called on object which does not have debug_mode attribute'
+        assert hasattr(
+            caller,
+            'logger'), 'skip_debug_mode called on object which has no logger'
+
+        if caller.debug_mode:
+            caller.logger.info(
+                "Skipping test in debug mode (requires release build)")
```

Comment:
To be usable on all tests, this needs to do a little extra work to alloc+free any nodes that haven't been used yet: otherwise tests fail on account of  the hygiene check for having asked for nodes they didn't use.

Not all tests need this (only ones that construct services during the test case), but to be broadly reusable this helper needs to handle it.

StorageResourcesTest.test_recovery_reads is an example of this: it does
```
        if self.debug_mode:
            self.logger.info("Skipping in debug mode")
            # Satisfy checks for test using all its nodes
            nodes = self.test_context.cluster.alloc(
                ClusterSpec.simple_linux(1))
            self.test_context.cluster.free_single(nodes[0])
            return
```
That just allocates 1 (special case for this test), I think it could be made general purpose by allocating self.test_context.cluster.available().size() nodes and then freeing them.

Reply:
makes sense, I have added a default cleanup action which can be overridden by defining `early_exit_hook` on the class.